### PR TITLE
Implement People You May Know.

### PR DESCRIPTION
**Pull request by yipingliao**
@andrewconner  @leogrim 

Implement front-end for People You May Know.

Wireframe linked in: https://app.asana.com/0/15014070792041/13515857251339 (you have to scroll down a bit to see it).

Still to do: when Leo's changes to the /friends/invite endpoint to change "users" to "kifi", I need to update that route.

### CODE REVIEW

File: kifi-backend/modules/shoebox/angular/src/friends/rightColFriendsView.js [hunk 1]
```diff
@@ -71,6 +71,121 @@ angular.module('kifi.friends.rightColFriendsView', [])
...
+          });
+        });
+        
+        var networkNamesMap = {
+          'facebook': 'Facebook',
+          'linkedin': 'LinkedIn',
+          'email': 'email'
+        };
+
+        if (peopleYouMayKnow.length < 3) {
```
Comment (yipingliao) [inline kifi-backend/modules/shoebox/angular/src/friends/rightColFriendsView.js:105]: If there aren't enough Kifi users to recommend, we grab people to invite.

File: kifi-backend/modules/shoebox/angular/src/invite/connectionCard.js [hunk 2]
```diff
@@ -74,7 +74,8 @@ angular.module('kifi.invite.connectionCard', ['angularMoment'])
         }
       } else {
         scope.invited = false;
-        scope.byline = network === 'email' ? inNetworkId : network.charAt(0).toUpperCase() + network.slice(1);
+        scope.byline = network === 'email' ? inNetworkId :
```
Comment (yipingliao) [inline kifi-backend/modules/shoebox/angular/src/invite/connectionCard.js:77]: Just a small change so that "LinkedIn" is displayed instead of "Linkedin".

File: kifi-backend/modules/shoebox/angular/src/friends/compactFriendsView.tpl.html [hunk 3]
```diff
@@ -1,9 +1,9 @@
-<div>
+<div ng-if="friends.length > 0">
```
Comment (yipingliao) [inline kifi-backend/modules/shoebox/angular/src/friends/compactFriendsView.tpl.html:1]: @andrewconner  this fixes that ugliness you saw earlier.

File: kifi-backend/modules/shoebox/angular/src/layout/rightCol/RightColCtrl.js [hunk 4]
```diff
@@ -22,9 +22,11 @@ angular.module('kifi.layout.rightCol', ['kifi.modal'])
       return profileService.me.seqNum > 0 && friendsReady;
     };
 
-    $scope.hasNoFriendsOrConnections = function () {
-      return (friendService.totalFriends() === 0) && (socialService.networks.length === 0);
-    };
+    socialService.refresh().then(function () {
```
Comment (yipingliao) [inline kifi-backend/modules/shoebox/angular/src/layout/rightCol/RightColCtrl.js:25]: This also is not related to People You May Know. I noticed a bug where we get the user's connected networks information incorrectly because we haven't refreshed socialService yet.

File: kifi-backend/modules/shoebox/angular/src/friends/rightColFriendsView.js [hunk 5]
```diff
@@ -71,6 +71,121 @@ angular.module('kifi.friends.rightColFriendsView', [])
...
+    link: function (scope/*, element, attrs*/) {
+      friendService.getPeopleYouMayKnow().then(function (people) {
+        var peopleYouMayKnow = [];
+
+        people.forEach(function (person) {
+          var name = person.firstName + ' ' + person.lastName;
+          peopleYouMayKnow.push({
+            id: person.id,
+            fullName: name + ', ',
+            pictureUrl: friendService.getPictureUrlForUser(person) || 'https://www.kifi.com/assets/img/ghost.100.png',
```
Comment (andrewconner) [inline kifi-backend/modules/shoebox/angular/src/friends/rightColFriendsView.js]: Hm perhaps we could improve `friendService.getPictureUrlForUser` to fall back to the ghost image.

Comment (yipingliao) [inline kifi-backend/modules/shoebox/angular/src/friends/rightColFriendsView.js]: Good idea. Will do.

Comment (yipingliao) [inline kifi-backend/modules/shoebox/angular/src/friends/rightColFriendsView.js]: Actually, nvm. I need to remove the fallback from here. I added it because the who to invite list needed it, but actually friendService.getPictureUrlForUser() always returns a non-empty url.

File: kifi-backend/modules/shoebox/angular/src/friends/rightColFriendsView.js [hunk 6]
```diff
@@ -71,6 +71,121 @@ angular.module('kifi.friends.rightColFriendsView', [])
...
+        });
+        
+        var networkNamesMap = {
+          'facebook': 'Facebook',
+          'linkedin': 'LinkedIn',
+          'email': 'email'
+        };
+
+        if (peopleYouMayKnow.length < 3) {
+          wtiService.getMore().then(function (wtiList) {
```
Comment (andrewconner) [inline kifi-backend/modules/shoebox/angular/src/friends/rightColFriendsView.js]: I believe this will cause wtiService to make an outgoing request every time the directive is rendered if `peopleYouMayKnow.length < 3`. ie, if `friendService.getPeopleYouMayKnow()` is too short, ask wtiService to give page 1. The next time we render, we ask wtiService to give page 2.

Since page 1 will always be more than 3 (if possible), you may want something like (not the perfect API, but best we can do without changing `wtiService` to know if it's been requested):

```
(wtiService.list.length > 0 ? $q.when(wtiService.list) : wtiService.getMore()).then(function () {
  // ... use wtiService.list
});
```

File: kifi-backend/modules/shoebox/angular/src/friends/rightColFriendsView.js [hunk 7]
```diff
@@ -71,6 +71,121 @@ angular.module('kifi.friends.rightColFriendsView', [])
...
 
+.directive('kfPeopleYouMayKnowView', 
+  ['$log', '$timeout', 'friendService', 'inviteService', 'wtiService', 
+  function ($log, $timeout, friendService, inviteService, wtiService) {
+  return {
+    replace: true,
+    restrict: 'A',
+    templateUrl: 'friends/peopleYouMayKnowView.tpl.html',
+    link: function (scope/*, element, attrs*/) {
+      friendService.getPeopleYouMayKnow().then(function (people) {
```
Comment (andrewconner) [inline kifi-backend/modules/shoebox/angular/src/friends/rightColFriendsView.js:82]: Since removal of a suggestion removes it from the page, it may make sense to do something like:

``` javascript
var pymk = friendService.getPeopleYouMayKnow();
var appendPeopleToSuggestions = function (people) {
 // ...
}
var wti = wtiService.list.length > 0 ? $q.when(wtiService.list) : wtiService.getMore();
var appendWtiToSuggestions = function () {
 // ... use wtiService.list
}
pymk.then(appendPeopleToSuggestions).then(appendWtiToSuggestions);
```

and show the box as soon as over 3 are populated (or both services return). This will make the list as long as possible, for the case of a user without tons of suggestions from us (common case), and they remove several people.

File: kifi-backend/modules/shoebox/angular/src/invite/connectionCard.js [hunk 8]
```diff
@@ -74,7 +74,8 @@ angular.module('kifi.invite.connectionCard', ['angularMoment'])
         }
       } else {
         scope.invited = false;
-        scope.byline = network === 'email' ? inNetworkId : network.charAt(0).toUpperCase() + network.slice(1);
+        scope.byline = network === 'email' ? inNetworkId : 
+          (network === 'linkedin' ? 'LinkedIn' : network.charAt(0).toUpperCase() + network.slice(1));
```
Comment (andrewconner) [inline kifi-backend/modules/shoebox/angular/src/invite/connectionCard.js:78]: Nice :)

**Comment by andrewconner** [comment]
Great job. Minor comments above, feel free to merge when you're ready.

File: kifi-backend/modules/shoebox/angular/src/friends/rightColFriendsView.js (hunk 7 above)
Comment (yipingliao) [inline kifi-backend/modules/shoebox/angular/src/friends/rightColFriendsView.js:82]: Not going to implement this because Leo's new endpoint will have both Kifi users and non-Kifi users so we will not be needing to ping wtiService when friendsService runs out.

File: kifi-backend/modules/shoebox/angular/src/friends/rightColFriendsView.js (hunk 6 above)
Comment (yipingliao) [inline kifi-backend/modules/shoebox/angular/src/friends/rightColFriendsView.js]: Good catch. Will do.
